### ISSUE TITLE
Server crash when many clients upload file in parallel over a TLS connection

### ISSUE THREAD

**Issue by amadio** [xrootd/xrootd#2226]
In trying to reproduce the client crash of the EOS probe, I got a crash on the server side:
```
Core was generated by `/usr/bin/xrootd -l /var/log/xrootd/xrootd.log -c /etc/xrootd/xrootd-secure.cfg'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  XrdLink::hasTLS (this=0x0) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/Xrd/XrdLink.hh:563
563	bool            hasTLS() const {return isTLS;}
[Current thread is 1 (Thread 0x7f70b3a006c0 (LWP 1092237))]
...
Thread 1 (Thread 0x7f70b3a006c0 (LWP 1092237)):
#0  XrdLink::hasTLS (this=0x0) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/Xrd/XrdLink.hh:563
#1  XrdXrootdProtocol::do_Protocol (this=this@entry=0x7f6f7401e030) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/XrdXrootd/XrdXrootdXeq.cc:1912
#2  0x00007f7136580078 in XrdXrootdProtocol::Process2 (this=this@entry=0x7f6f7401e030) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/XrdXrootd/XrdXrootdProtocol.cc:471
#3  0x00007f7136580788 in XrdXrootdProtocol::Process (this=0x7f6f7401e030, lp=<optimized out>) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/XrdXrootd/XrdXrootdProtocol.cc:430
#4  0x00007f71364ba84e in XrdLinkXeq::DoIt (this=0x7f71200078b8) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/Xrd/XrdLinkXeq.cc:320
#5  0x00007f71364be0c5 in XrdScheduler::Run (this=0x556dfb3c4720 <XrdGlobal::Sched>) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/Xrd/XrdScheduler.cc:406
#6  0x00007f71364be142 in XrdStartWorking (carg=<optimized out>) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/Xrd/XrdScheduler.cc:89
#7  0x00007f713644d14d in XrdSysThread_Xeq (myargs=0x7f6ee4013060) at /usr/src/debug/net-libs/xrootd-5.6.9/xrootd-5.6.9/src/XrdSys/XrdSysPthread.cc:86
#8  0x00007f7135eb2cc6 in start_thread (arg=<optimized out>) at pthread_create.c:447
#9  0x00007f7135f2619c in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```
This happened while running
```
while [[ $? == 0 ]]; do for i in {0..64}; do env XrdSecPROTOCOL=gsi LD_LIBRARY_PATH=$PWD/src/XrdCl ./src/XrdCl/xrdcp -S 15 -f file${i}.raw roots://gentoo.cern.ch//file${i}.raw & done; wait; done
```
to upload 64 zero-size files in parallel over TLS to a server. I just did `touch file{0..64}.raw` before running the above. The full server configuration is:
```
all.trace    all
http.trace   all
tpc.trace    all
xrd.trace    all
xrootd.trace all
scitokens.trace all

all.export /
all.sitename  XRootD
all.adminpath /tmp
oss.localroot /srv/xrootd
xrootd.chksum chkcgi adler32 crc32c
xrootd.seclib libXrdSec.so

xrd.tlsca certdir /etc/grid-security/certificates
xrd.tls /etc/xrootd/hostcert.pem /etc/xrootd/hostkey.pem

sec.protocol krb5 /etc/xrootd/krb5.keytab host/gentoo.cern.ch@CERN.CH
sec.protocol sss -c /etc/xrootd/xrootd.keytab -s /etc/xrootd/xrootd.keytab
sec.protocol gsi -cert:/etc/xrootd/hostcert.pem -key:/etc/xrootd/hostkey.pem -gridmap:/etc/xrootd/grid-mapfile
# sec.protocol ztn -tokenlib libXrdAccSciTokens.so
# sec.protbind * only krb5 sss ztn gsi
sec.protbind * only krb5 gsi sss

xrd.port 1094
xrd.protocol XrdHttp:8080 libXrdHttp.so
http.exthandler xrdtpc libXrdHttpTPC.so
http.secxtractor libXrdSecgsiVOMS.so

xrd.protocol XrdHttp:8080 libXrdHttp.so
http.header2cgi Authorization authz
```

**Comment by abh3** [comment]
Hopefully there is a core file here. I can't see how the link pointer could become zero. If this is easily reproducible we have a chance of finding out.

**Comment by amadio** [comment]
Hi @abh3, yes, I have a core file, but for Gentoo. I will try it on CentOS or Alma and send you the core file. What OS are you using? Alma 9?

**Comment by abh3** [comment]
Hi Guilherme,

Sending me the core file is unlikely to be helpful. Once typically needs to look at it in the same environment that created it.

Andy

________________________________

**Comment by amadio** [comment]
I'm trying to revisit this issue. I managed to get a crash again, but with a different stack trace than the one in the description:
```
Core was generated by `/usr/bin/xrootd -l /var/log/xrootd/xrootd.log -c /etc/xrootd/xrootd-secure.cfg -k fifo -s /run/xrootd/xrootd-secure.pid -n secure'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f1fd10b3f69 in XrdXrootdProtocol::do_Protocol (this=this@entry=0x7f1d1c020c80)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/XrdXrootd/XrdXrootdXeq.cc:1964
1964               eDest.Emsg("Xeq", "Unable to enable TLS for", Link->ID);
[Current thread is 1 (Thread 0x7f1ca3fff6c0 (LWP 311031))]
(gdb) bt
#0  0x00007f1fd10b3f69 in XrdXrootdProtocol::do_Protocol (this=this@entry=0x7f1d1c020c80)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/XrdXrootd/XrdXrootdXeq.cc:1964
#1  0x00007f1fd10abf4c in XrdXrootdProtocol::Process2 (this=this@entry=0x7f1d1c020c80)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/XrdXrootd/XrdXrootdProtocol.cc:471
#2  0x00007f1fd10ac683 in XrdXrootdProtocol::Process (this=0x7f1d1c020c80, lp=<optimized out>)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/XrdXrootd/XrdXrootdProtocol.cc:430
#3  0x000055cde9d4ede2 in XrdProtLoad::Process (this=0x7ffee072d0b0, lp=0x7f1fc400a020)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdProtLoad.cc:236
#4  0x00007f1fd0fdefa6 in XrdLinkXeq::DoIt (this=0x7f1fc400a020)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdLinkXeq.cc:304
#5  0x00007f1fd0fdb79b in XrdLink::setProtocol (this=0x7f1fc400a020, pp=<optimized out>, runit=runit@entry=true,
    push=push@entry=false) at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdLink.cc:435
#6  0x000055cde9d5066c in XrdMain::DoIt (this=0x7ffee072d120)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdMain.cc:99
#7  0x00007f1fd0fe2c29 in XrdScheduler::Run (this=0x55cde9d56120 <XrdGlobal::Sched>)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdScheduler.cc:365
#8  0x00007f1fd0fe2caa in XrdStartWorking (carg=<optimized out>)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/Xrd/XrdScheduler.cc:89
#9  0x00007f1fd0f6c77f in XrdSysThread_Xeq (myargs=0x7f1bc4001090)
    at /usr/src/debug/net-libs/xrootd-9999/xrootd-9999/src/XrdSys/XrdSysPthread.cc:86
#10 0x00007f1fd0abed7a in start_thread (arg=<optimized out>) at pthread_create.c:448
#11 0x00007f1fd0b274ec in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```
This is using the current `master` branch.

**Comment by smithdh** [comment]
Hi. In some tests I was no longer able to reproduce a server crash if I made this change:

```
diff --git a/src/Xrd/XrdPollE.icc b/src/Xrd/XrdPollE.icc
index e6b11d24e..ea1869558 100644
--- a/src/Xrd/XrdPollE.icc
+++ b/src/Xrd/XrdPollE.icc
@@ -328,7 +328,7 @@ void XrdPollE::Start(XrdSysSemaphore *syncsem, int &retcode)
            {if (PollTab[i].data.ptr == &WaitFd)
               {haveWaiters = true; waitFdEvents = PollTab[i].events;}
             else if ((pInfo = (XrdPollInfo *)PollTab[i].data.ptr))
-              {if (!(pInfo->isEnabled) && pInfo->FD >= 0)
+              {if (!pInfo->isEnabled)
                   remFD(*pInfo, PollTab[i].events);
                   else {pInfo->isEnabled = 0;
                         if (!(PollTab[i].events & pollOK)
```

but this is effectively reversing commit ac6e37f ["[Server] Avoid misleading error message due to queued but delayed event."]. @abh3 do you remember the error message which was being avoided?

**Comment by smithdh** [comment]
I been looking at this further: I don't think the change above is a appropriate fix (i.e. my observation about (pInfo->FD>=0), altough it was a hint. Currently I think there are probably two bugs contributing to this issue:

(1) XrdPollE::Wait4Poller() does not completely avoid the situation and it was designed to fix. (See comment in XrdPollE::Exclude() for what it should avoid). Hopefully this should have a relative simple & localised fix. This may fix many instances of the crash in this ticket (#2226), and may have a small role in #2602. (However I believe there's another issue to be fixed for #2602).

(2) The bound stream close (in XrdXrootdProtocol::Cleanup(), XrdXrootdProtocol.cc:1062) does not seem completely interlocked with possible close from XrdLinkXeq::Close(), which should also be relevant to this ticket.

I have a possible fix for (i), and maybe even (ii), but the latter appears to need a more wide-ranging change. Unless Andy prefers to spend time on this, I could eventually open a PR for review when I have something. But we'll likely to have iterate on it a bit so that I don't break any intended design for XrdLink. (And if it's to be included in 5.9.x, that also limits what can be changed).
